### PR TITLE
Database set tree

### PR DESCRIPTION
**Pull request by blairsteven**
Set values in the database via a bulk GNode operation rather than path by path.

### CODE REVIEW

**Review by sparlane — APPROVED** [review]

**Review by carlgsmith — APPROVED** [review]

File: test.c [hunk 1]
```diff
@@ -955,17 +955,28 @@ _indexer_writes (const char *d)
...
 void
 test_index_writes ()
 {
     char *path = TEST_PATH"/counters/*";
+    char *provide_path = TEST_PATH"/counters/*/something/else";
     GNode *root;
 
     CU_ASSERT (apteryx_index (path, _indexer_writes));
-    CU_ASSERT ((root = apteryx_get_tree (TEST_PATH"/counters")) == NULL);
+    /* There's no reason to call the indexer without a provider behind it. */
```
Comment (carlgsmith) [inline test.c:972]: Do we have a test for a leaflist e.g. TEST_PATH"/users" where it is a list of usernames only - in other words there is no provider and the indexer is providing all the data?

File: test.c [hunk 2]
```diff
@@ -5012,32 +5172,83 @@ test_perf_get_tree_real ()
...
+    printf ("%"PRIu64"us(%"PRIu64"us) ... ", time, time/(n));
     apteryx_free_tree (root);
 exit:
     CU_ASSERT (apteryx_prune (path));
     CU_ASSERT (assert_apteryx_empty ());
 }
 
+
+void
+test_perf_get_tree_5000a ()
```
Comment (carlgsmith) [inline test.c:5206]: 5000a?

File: rpc.c [hunk 3]
```diff
@@ -758,15 +758,41 @@ _rpc_msg_decode_tree (rpc_message msg, GNode *root)
 
                     /* Actually create the root node. */
                     root = APTERYX_NODE (NULL, path);
+
+                    /* Add null value - this may be superceded later */
+                    g_node_prepend_data(root, NULL);
```
Comment (carlgsmith) [inline rpc.c:763]: Is there a cleaner way of doing this - its a bit ugly to have to remove the node below.

File: callbacks.c [hunk 4]
```diff
@@ -215,7 +215,9 @@ cb_gather_search (struct callback_node *node, GList *callbacks_so_far, const cha
         for (GList *iter = children; iter; iter = iter->next)
         {
             struct callback_node *child = iter->data;
-            if (g_list_length (child->exact) > 0 || !hashtree_empty (&child->hashtree_node))
+            if (g_list_length (child->exact) > 0 ||
+                g_list_length (child->directory) > 0 ||
```
Comment (carlgsmith) [inline callbacks.c:219]: I think checking child->exact and child->directory pointers is enough here.

File: apteryx.c [hunk 5]
```diff
@@ -1250,44 +1247,56 @@ apteryx_get_tree (const char *path)
...
-        DEBUG ("   = %s\n", value);
-        g_node_append_data (root, (gpointer) strdup (value));
-    }
-    else if (path)
+
+    /* Read tree from client */
+    root = rpc_msg_decode_tree (&msg);
+
+    /* rpc_msg_decode_tree compresses the trunk of the tree to be as short as possible,
+     * reset the root node key here to match the request */
```
Comment (carlgsmith) [inline apteryx.c:1255]: I wonder if this can be optimised

File: Makefile [hunk 6]
```diff
@@ -54,7 +54,7 @@ libapteryx.so: libapteryx.so.$(ABI_VERSION)
 	@echo "Compiling "$<""
 	$(Q)$(CC) $(CFLAGS) $(EXTRA_CFLAGS) -c $< -o $@
 
-apteryxd: apteryxd.c hashtree.c database.c rpc.o rpc_transport.o rpc_socket.o config.o callbacks.o
+apteryxd: apteryxd.c hashtree.c database.c rpc.o rpc_transport.o rpc_socket.o config.o callbacks.o libapteryx.so
```
Comment (carlgsmith) [inline Makefile:57]: huh?

**Review by sparlane — APPROVED** [review]

**Pull request merged**